Update Assurance Object - Verification Method Object (#2169)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/api/policy/index.md

| `constraints`        | Array of [Constraint Object](#constraints)           | A JSON array containing nested authenticator constraint objects which are organized by Authenticator Class.                                                                           | [Constraint Object](#constraints) consisting of `POSSESSION` constraint or `KNOWLEDGE` constraint or both. See [Verification Method JSON Examples](#verification-method-JSON-examples)                                                                      |
| `reauthenticateIn`   | String (ISO 8601) | The period after which the end-user must be reauthenticated, regardless of activity.                                  | N/A                                                                                               |
| `inactivityPeriod`   | String (ISO 8601) | The inactivity time period after which the End User must re-authenticate                                                | N/A                                                                                               |

#### Constraints

> ##### Constrints default example

Reviewer: ```suggestion
##### Constraints default example
```

| `hardwareProtection` | String            | Indicates if any secrets or private keys that are used during authentication must be hardware protected and not exportable. This property is only set for `POSSESSION` constraints.| `REQUIRED`, `OPTIONAL`                                                                            |
| `deviceBound` | String            | Indicates if device-bound factors are required. This property is only set for  `POSSESSION` constraints. | `REQUIRED`, `OPTIONAL`                                                                            |
| `phishingResistant` | String            | Indicates if phishing-resistant factors are required. This property is only set for `POSSESSION` constraints. | `REQUIRED`, `OPTIONAL`                                                                            |
| `userPresence` | String            | Indicates whether the user needs to approve a prompt in Okta Verify or provide biometrics (meets NIST AAL2 requirements). This property can be set in `POSSESSION` constraint only.| `REQUIRED`, `OPTIONAL`                                                                            |

> | `reauthenticateIn`   | String (ISO 8601) | The time that the End User must re-authenticate, regardless of user activity                            | N/A                                                                                               |

Reviewer 2: This table should have a `Default` column, because `userPresence` defaults to `REQUIRED` and the others default to `OPTIONAL`.

| `reauthenticateIn`   | String (ISO 8601) | The time that the End User must re-authenticate, regardless of user activity                                  | N/A                                                                                               |
| `inactivityPeriod`   | String (ISO 8601) | The inactivity time period after which the End User must re-authenticate                                                | N/A                                                                                               |

#### Constraints

Each nested constraint object is treated as a list, all of which must be satisfied. The top-level array is treated as a set, one of which must be satisfied.

##### Constraints default example
```json
"constraints": [
  { // object 1
    "knowledge": {  // 1A
      "types": [
        "password"
      ],
      "reauthenticateIn": "PTOS"
    },
    "possession": { // 1B
      "userPresence": "OPTIONAL"
    }
  },
  { // object 2
    "knowledge": {  // 2A 
      "types": [
        "password"
      ],
    },
    "possession": { // 2B
      "phishingResistant": "REQUIRED"
    }
  }
]

> ```

Author: Rendering of the JSON object ![Screen Shot 2021-05-27 at 10 30 44 AM](https://user-images.githubusercontent.com/48875603/119871018-de7dd780-bed6-11eb-8011-44f30c016199.png)

In the above example contraint object 1 (password factor with reauthenticatation on every signin attempt and a possession factor) or constraint object 2 (password factor  and a possession factor which is a phishing resistant factor liked WebAuthn ) must be used to satisfy the assurance policy.

This can be read logically as: `( (1A && 1B) || (2A && 2B) )`

The number of authenticator class constraints in each constraint object must be less than or equal to the value of `factorMode`. If the value is less, there are no constraints on any additional factors.

| Property            | Data Type              | Description                                                                                                             | Supported Values                                  | Default |                                                
| -------------------- | ----------------- | ----------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------- |-----------|

> | `types`              | Array  of Authenticator Types           | The Authenticator Types that are permitted                                                                           | [ `SECURITY_KEY`, `PHONE`, `EMAIL`, `PASSWORD`, `SECURITY_QUESTION`, `APP`, `FEDERATED` ]                         |  N/A|

Reviewer 3: ```suggestion
| `types`              | Array  of Authenticator types           | The Authenticator types that are permitted.                                                                           | [ `SECURITY_KEY`, `PHONE`, `EMAIL`, `PASSWORD`, `SECURITY_QUESTION`, `APP`, `FEDERATED` ]                         |  N/A|
```

> | `methods`            | Array of Authenticator Methods           | The Authenticator Methods that are permitted                                                                          | [ `PASSWORD`, `SECURITY_QUESTION`, `SMS`, `VOICE`, `EMAIL`, `FIDO2`, `PUSH`, `SIGNED_NONCE`, `OTP`, `WEBAUTHN`, `DUO`, `IDP` ] |  N/A|

Reviewer 3: ```suggestion
| `methods`            | Array of Authenticator methods           | The Authenticator methods that are permitted.                                                                          | [ `PASSWORD`, `SECURITY_QUESTION`, `SMS`, `VOICE`, `EMAIL`, `FIDO2`, `PUSH`, `SIGNED_NONCE`, `OTP`, `WEBAUTHN`, `DUO`, `IDP` ] |  N/A|
```

| `hardwareProtection` | String            | Indicates if any secrets or private keys that are used during authentication must be hardware protected and not exportable. This property is only set for `POSSESSION` constraints.| `REQUIRED`, `OPTIONAL`     |   `OPTIONAL`|
| `deviceBound` | String            | Indicates if device-bound factors are required. This property is only set for `POSSESSION` constraints. | `REQUIRED`, `OPTIONAL`                                                                            |`OPTIONAL`|
| `phishingResistant` | String            | Indicates if phishing-resistant factors are required. This property is only set for `POSSESSION` constraints. | `REQUIRED`, `OPTIONAL`                                                                            |`OPTIONAL`|

> | `userPresence` | String            | Indicates whether the user needs to approve a prompt in Okta Verify or provide biometrics (meets NIST AAL2 requirements).  This property is only set for `POSSESSION` constraints.| `REQUIRED`, `OPTIONAL`                                                                            |`REQUIRED`|

Reviewer 3: ```suggestion
| `userPresence` | String            | Indicates if the user needs to approve an Okta Verify prompt or provide biometrics (must meet NIST AAL2 requirements). This property is only set for `POSSESSION` constraints.| `REQUIRED`, `OPTIONAL`                                                                            |`REQUIRED`|
```

> | `reauthenticateIn`   | String (ISO 8601) | The time that the End User must re-authenticate, regardless of user activity                            | N/A                                                                                               | N/A|

Reviewer 3: ```suggestion
| `reauthenticateIn`   | String (ISO 8601) | The duration that the End User must re-authenticate, regardless of user activity. Use the ISO 8601 Period format for recurring time intervals.                            | `P<date>T<time>`                                                                                               | N/A|
```

| `factorMode`         | String            | The number of factors required to satisfy this assurance level.                                                         | `1FA`, `2FA`                                                                                      |
| `type`         | String            | Verification method type.                                                         | `ASSURANCE`     |
| `constraints`        | [Constraint Object](#constraints)           | A JSON array containing nested authenticator constraint objects which are organized by Authenticator Class.                                                                           | `POSSESSION`, `KNOWLEDGE`                                                                         |
| `reauthenticateIn`   | String (ISO 8601) | The period after which the end-user should be reauthenticated, regardless of activity.                                  | N/A                                                                                               |

> | `inactivityPeriod`   | String (ISO 8601) | The period of inactivity after which the user should be reauthenticated.                                                | N/A                                                                                               |

Author: @antonioruberto-okta Do you know if `inactivityPeriod` is still relevant? I don't see any usage for this property.

#### Constraints
Each nested constraint object is treated as a list, all of which must be satisfied. The top-level array is treated as a set, one of which must be satisfied.

This can be read logically as: `( (1A && 1B) || (2A && 2B) )`

The number of authenticator class constraints in each constraint object be less than or equal to the value of `factorMode`. If the value is less, there are no constraints on any additional factors.

| Property            | Data Type              | Description                                                                                                             | Supported Values                                                                                  |
| -------------------- | ----------------- | ----------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------- |

> | `types`              | Array             | The Authenticator Types which are permitted.                                                                            | `SECURITY_KEY`, `PHONE`, `EMAIL`, `PASSWORD`, `SECURITY_QUESTION`                          |

Reviewer 4: Are these the only types supported? type `app`?

| `methods`            | Array             | The Authenticator Methods which are permitted.                                                                          | `PASSWORD`, `SECURITY_QUESTION`, `SMS`, `VOICE`, `EMAIL`, `FIDO2` |

> | `hardwareProtection` | String            | Indicates whether any secrets or private keys used during authentication must be hardware protected and not exportable. | `REQUIRED`, `OPTIONAL`                                                                            |

Reviewer 4: Do we also need to doc 'phishingResistant', 'deviceBound' or 'userVerification'

Reviewer 4: Also, do we need to mention that these are possession constraints?

> | `reauthenticateIn`   | String (ISO 8601) | The period after which the end-user should be reauthenticated, regardless of activity.                                  | N/A                                                                                               |

Reviewer 5: `must be` ?

Multi-factor authentication (MFA) is the use of more than one factor. MFA is the most common way to increase assurance. Authenticators also have other characteristics that may raise or lower assurance. For example, possession factors may be implemented in software or hardware, with hardware being able to provide greater protection when storing shared secrets or private keys, and thus providing higher assurance.

> | Parameter            | Type              | Description                                                                                                             | Supported Values                                                                                  |

Author: Not adding a OIE only note for `Verification Method Object` and `Constraints` objects, since the parent object `#App-Sign-On-Polic`y  has it.

| -------------------- | ----------------- | ----------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------- |
| `factorMode`         | String            | The number of factors required to satisfy this assurance level.                                                         | `1FA`, `2FA`                                                                                      |
| `type`         | String            | Verification method type.                                                         | `ASSURANCE`     |

> | `constraints`        | [Constraint Object](#constraints)           | A JSON array containing nested authenticator constraint objects which are organized by Authenticator Class.                                                                           | `POSSESSION`, `KNOWLEDGE`                                                                         |

Reviewer 6: The supported values look wrong here. Also Data type maybe should be `Array of [Constraint Object](#constraints)`?

> | `constraints`        | [Constraint Object](#constraints)           | A JSON array containing nested authenticator constraint objects which are organized by Authenticator Class.                                                                           | `POSSESSION`, `KNOWLEDGE`                                                                         |

Author: Addresses it locally and pushed because of a conflict.

| `reauthenticateIn`   | String (ISO 8601) | The period after which the end-user must be reauthenticated, regardless of activity.                                  | N/A                                                                                               |

> | `inactivityPeriod`   | String (ISO 8601) | The period of inactivity after which the user must be reauthenticated.                                                | N/A                                                                                               |

Reviewer 2: Didn't we say that this doesn't actually do anything right now?  Did we decide to leave it there because it's already there?

Author: Not sure what has happened here and why it's showing as a property that's added. https://github.com/okta/okta-developer-docs/pull/2169#discussion_r639012239. Created https://oktainc.atlassian.net/browse/OKTA-399402 to  investigate further.

#### Constraints
Each nested constraint object is treated as a list, all of which must be satisfied. The top-level array is treated as a set, one of which must be satisfied.

> This can be read logically as: `( (1A && 1B) || (2A && 2B) )`

Reviewer 2: What is 'This' referring to here?  Is there a diagram or something else on the rendered doc page that references `1A`, `1B` etc...?

Author: Added a JSON example explaining what 1A, 1B, 2A, 2B mean.

> The number of authenticator class constraints in each constraint object be less than or equal to the value of `factorMode`. If the value is less, there are no constraints on any additional factors.

Reviewer 2: Think this should be "The number of authenticator class constraints in each constraint object **must** be less than or equal to..."  (missing 'must').

| Property            | Data Type              | Description                                                                                                             | Supported Values                                                                                  |
| -------------------- | ----------------- | ----------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------- |
| `types`              | Array             | The Authenticator Types which are permitted.                                                                            | `SECURITY_KEY`, `PHONE`, `EMAIL`, `PASSWORD`, `SECURITY_QUESTION`, `APP`, `FEDERATED`                          |

> | `methods`            | Array             | The Authenticator Methods which are permitted.                                                                          | `PASSWORD`, `SECURITY_QUESTION`, `SMS`, `VOICE`, `EMAIL`, `FIDO2`, `PUSH`, `SIGNED_NONCE`, `OTP`, `WEBAUTHN`, `DUO`, `IDP` |

Reviewer 3: ```suggestion
| `methods`            | Array of Authenticator Methods           | The Authenticator Methods that are permitted                                                                          | [ `PASSWORD`, `SECURITY_QUESTION`, `SMS`, `VOICE`, `EMAIL`, `FIDO2`, `PUSH`, `SIGNED_NONCE`, `OTP`, `WEBAUTHN`, `DUO`, `IDP` ] |
```

Reviewer 3: Added "Array of Authenticator Methods" and square brackets around the array. Please remove if these are not appropriate. Thanks

> | `hardwareProtection` | String            | Indicates whether any secrets or private keys used during authentication must be hardware protected and not exportable. This property can be set in `POSSESSION` constraint only.| `REQUIRED`, `OPTIONAL`                                                                            |

Reviewer 3: ```suggestion
| `hardwareProtection` | String            | Indicates if any secrets or private keys that are used during authentication must be hardware protected and not exportable. This property is only set for `POSSESSION` constraints.| `REQUIRED`, `OPTIONAL`                                                                            |
```

> | `deviceBound` | String            | Indicates whether device bound factors are required. This property can be set in `POSSESSION` constraint only. | `REQUIRED`, `OPTIONAL`                                                                            |

Reviewer 3: ```suggestion
| `deviceBound` | String            | Indicates if device-bound factors are required. This property is only set for  `POSSESSION` constraints. | `REQUIRED`, `OPTIONAL`                                                                            |
```

> | `phishingResistant` | String            | Indicates whether phishing resistant factors are required. This property can be set in `POSSESSION` constraint only. | `REQUIRED`, `OPTIONAL`                                                                            |

Reviewer 3: ```suggestion
| `phishingResistant` | String            | Indicates if phishing-resistant factors are required. This property is only set for `POSSESSION` constraints. | `REQUIRED`, `OPTIONAL`                                                                            |
```

> | `userPresence` | String            | Indicates if user presence is required or not.  This property can be set in `POSSESSION` constraint only.| `REQUIRED`, `OPTIONAL`                                                                            |

Reviewer 2: Did the descriptions of these properties come from someone on the dev docs team?  I wonder if we should try to better explain what they mean.

For example, the UI option for `userPresence` says something like "Whether the user is required to accept a consent prompt" or something like that.

> | `userPresence` | String            | Indicates if user presence is required or not.  This property can be set in `POSSESSION` constraint only.| `REQUIRED`, `OPTIONAL`                                                                            |

Reviewer 3: ```suggestion
| `userPresence` | String            | Indicates if user presence is required.  This property is only set for `POSSESSION` constraints.| `REQUIRED`, `OPTIONAL`                                                                            |
```

Author: Addresses it locally and pushed because of a conflict.

> | `reauthenticateIn`   | String (ISO 8601) | The period after which the end-user must be reauthenticated, regardless of activity.                                  | N/A                                                                                               |

Reviewer 3: ```suggestion
| `reauthenticateIn`   | String (ISO 8601) | The time that the End User must re-authenticate, regardless of user activity                            | N/A                                                                                               |
```

Reviewer 3: Not sure if this is an interval or an exact time? Use "time interval" if it's an interval, otherwise, use "time".
